### PR TITLE
rust: bindgen fixes

### PR DESCRIPTION
**Pull request by K900**
###### Description of changes

Basically:
- we used to build rustfmt as if it comes from nightly Rust, which is unexpected and therefore bad
- the failing test in rustfmt is marked as nightly-only, even though it doesn't actually require nightly
- said test is also actually broken, which is bad
- building rustfmt as stable "fixed" it by not running the test at all, which is fine
- bindgen's tests require a rustfmt with unstable features, which is also not great
- so fix the test properly and add a second rustfmt-nightly that's built as if it's nightly, and make bindgen use that

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/rust/rustfmt.nix [hunk 1]
```diff
@@ -1,7 +1,7 @@
-{ lib, stdenv, rustPlatform, Security }:
+{ lib, stdenv, rustPlatform, Security, asNightly ? false }:
 
 rustPlatform.buildRustPackage rec {
-  pname = "rustfmt";
+  pname = if asNightly then "rustfmt-nightly" else "rustfmt";
```
Comment (06kellyjac) [inline pkgs/development/compilers/rust/rustfmt.nix:4]:
```suggestion
  pname = "rustfmt" + lib.optionalString asNightly "-nightly";
```

?

Comment (K900) [reply pkgs/development/compilers/rust/rustfmt.nix:4]: No preference either way, but sure.

File: pkgs/development/compilers/rust/rustfmt.nix [hunk 2]
```diff
@@ -21,6 +21,8 @@ rustPlatform.buildRustPackage rec {
   CFG_RELEASE = rustPlatform.rust.rustc.version;
   CFG_RELEASE_CHANNEL = "stable";
 
+  patches = [ ./rustfmt-fix-self-tests.patch ];
```
Comment (zowoq) [inline pkgs/development/compilers/rust/rustfmt.nix:24]: Can we use `fetchpatch` here rather than committing the patch?

Comment (K900) [reply pkgs/development/compilers/rust/rustfmt.nix:24]: Not really, upstream rustc just does big merges from the rustfmt repo from time to time, so there's no individual commit with a fix we can pick.

Comment (zowoq) [reply pkgs/development/compilers/rust/rustfmt.nix:24]: Should be able to get it from `rustfmt` and adjust it to match with `extraPrefix`.

Comment (K900) [reply pkgs/development/compilers/rust/rustfmt.nix:24]: https://github.com/rust-lang/rustfmt/commit/89ca3f3a100456d652b156b1a62b8e244e526c4e

What is even.

Comment (06kellyjac) [reply pkgs/development/compilers/rust/rustfmt.nix:24]: It's a pretty basic change. I'm not sure its worth the effort to do that

Comment (zowoq) [reply pkgs/development/compilers/rust/rustfmt.nix:24]: Okay, upstream patch is no good. Please add a comment instead. "fixed upstream, can be removed next release" or similar.

File: pkgs/development/compilers/rust/default.nix [hunk 3]
```diff
@@ -77,6 +77,7 @@ in
         pkgsBuildTarget = pkgsBuildTarget // { targetPackages.stdenv = llvmBootstrapForDarwin.stdenv; };
       });
       rustfmt = self.callPackage ./rustfmt.nix { inherit Security; };
+      rustfmt-nightly = self.callPackage ./rustfmt.nix { inherit Security; asNightly = true; };
```
Comment (Artturin) [inline pkgs/development/compilers/rust/default.nix:80]: Someone could think that this is actually nightly

Maybe `rustfmt-pose-as-nightly`

Comment (K900) [reply pkgs/development/compilers/rust/default.nix:80]: Yeah I don't like the name too much either. Maybe `rustfmt-as-nightly`?

Comment (Artturin) [reply pkgs/development/compilers/rust/default.nix:80]: rustfmt-fake-nightly?

Comment (K900) [reply pkgs/development/compilers/rust/default.nix:80]: Here's an idea: just drop the alias altogether, and have bindgen depend on `rustfmt.override { asNightly = true; }`.

**Review by alyssais — COMMENTED** [review]
LGTM with the `rustfmt-nightly` alias dropped.

**Comment by K900** [comment]
Aight, alias dropped.

**Comment by Artturin** [comment]
ofborg builds but the test shows
```
Failed to run rustfmt: cannot find binary path (non-fatal, continuing)
```

**Comment by K900** [comment]
Yeah, that's fine. Bindgen will try to reformat the generated code with the user's Rustfmt, but we don't care, since there's no user.

**Pull request merged**